From 7d36f24cd9fef0678d8b1bdf9bb375c8c055259f Mon Sep 17 00:00:00 2001
From: Ankit Jain <ankitja@vmware.com>
Date: Tue, 21 Feb 2023 12:35:17 +0000
Subject: [PATCH] cqlshlib: Added support for python-3.11

File "/usr/lib/python3.11/site-packages/cqlshlib/saferscanner.py", line 73, in __init__
            s = re.sre_parse.State()
                ^^^^^^^^^^^^
AttributeError: module 're' has no attribute 'sre_parse'

Fix:
- Ref: https://github.com/apache/cassandra/commit/0b5248d8c3ef743a8dda42f7be5dadb70e373b51

Signed-off-by: Ankit Jain <ankitja@vmware.com>
---
 cqlshlib/saferscanner.py | 26 ++++++++++++++++++++++++--
 1 file changed, 24 insertions(+), 2 deletions(-)

diff --git a/cqlshlib/saferscanner.py b/cqlshlib/saferscanner.py
index 3cc3430..b0922c3 100644
--- a/cqlshlib/saferscanner.py
+++ b/cqlshlib/saferscanner.py
@@ -19,7 +19,10 @@
 # regex in-pattern flags. Any of those can break correct operation of Scanner.
 
 import re
-from sre_constants import BRANCH, SUBPATTERN, GROUPREF, GROUPREF_IGNORE, GROUPREF_EXISTS
+try:
+    from sre_constants import BRANCH, SUBPATTERN, GROUPREF, GROUPREF_IGNORE, GROUPREF_EXISTS
+except ImportError:
+    from re._constants import BRANCH, SUBPATTERN, GROUPREF, GROUPREF_IGNORE, GROUPREF_EXISTS
 from sys import version_info
 
 
@@ -80,5 +83,24 @@ class Py38SaferScanner(SaferScannerBase):
         self.p = p
         self.scanner = re.sre_compile.compile(p)
 
+class Py311SaferScanner(SaferScannerBase):
 
-SaferScanner = Py38SaferScanner if version_info >= (3, 8) else Py36SaferScanner
+    def __init__(self, lexicon, flags=0):
+        self.lexicon = lexicon
+        p = []
+        s = re._parser.State()
+        s.flags = flags
+        for phrase, action in lexicon:
+            gid = s.opengroup()
+            p.append(re._parser.SubPattern(s, [(SUBPATTERN, (gid, 0, 0, re._parser.parse(phrase, flags))), ]))
+            s.closegroup(gid, p[-1])
+        p = re._parser.SubPattern(s, [(BRANCH, (None, p))])
+        self.p = p
+        self.scanner = re._compiler.compile(p)
+
+if version_info >= (3, 11):
+    SaferScanner = Py311SaferScanner
+elif version_info >= (3, 8):
+    SaferScanner = Py38SaferScanner
+else:
+    SaferScanner = Py36SaferScanner
-- 
2.23.1

